Userland/Libraries/LibWeb/WebAudio/DynamicsCompressorNode.cpp

```diff
@@ -23,6 +23,12 @@ WebIDL::ExceptionOr<JS::NonnullGCPtr<DynamicsCompressorNode>> DynamicsCompressor
 // https://webaudio.github.io/web-audio-api/#dom-dynamicscompressornode-dynamicscompressornode
 WebIDL::ExceptionOr<JS::NonnullGCPtr<DynamicsCompressorNode>> DynamicsCompressorNode::construct_impl(JS::Realm& realm, JS::NonnullGCPtr<BaseAudioContext> context, DynamicsCompressorOptions const& options)
 {
+    // Validate the channel_count_mode before creating the node
+    if (options.channel_count_mode == Bindings::ChannelCountMode::Max) {
+        // Return a NotSupportedError if 'max' is used
```

Comment:
Where in the spec is this constructor behavior mentioned?

Reply:
https://webaudio.github.io/web-audio-api/#dom-audionode-channelcountmode
```
The channel count mode cannot be set to "[max](https://webaudio.github.io/web-audio-api/#dom-channelcountmode-max)", and a [NotSupportedError](https://webidl.spec.whatwg.org/#notsupportederror) exception MUST be thrown for any attempt to set it to "[max](https://webaudio.github.io/web-audio-api/#dom-channelcountmode-max)".
```

Reply:
I'd copy that passage verbatim to the setter, and call the setter here instead.

Reply:
The problem is that I cannot throw NotSupportedError from the constructor (DynamicsCompressorNode::DynamicsCompressorNode one), that's why there's a separate check in DynamicsCompressorNode::construct_impl
Setter (set_channel_count_mode) already checks that the value is not Max. However, I have not found a way to use that

Reply:
You create an instance of DynamicsCompressorNode in construct_impl, you can call the setter after that?

Reply:
In the spec, this is meant to be done by: 'FIXME: Invoke "Initialize the AudioNode" steps.', if I am understanding correctly, which is common for all of the AudioNodes.

Specifically - step two of that algorithm -

> 2. Set its value for numberOfInputs, numberOfOutputs, channelCount, channelCountMode, channelInterpretation to the default value for this specific interface outlined in the section for each AudioNode."

I would recommend that we change `AudioNodeOptions` so that `channel_count_mode` and `channel_interpretation` are both `Optional`. What we can then do is have a common helper function within AudioNode which will set the audio options for a given AudioNode.

It could look something like:

```
ExceptionOr<void> set_audio_node_options(AudioNodeOptions const&, DefaultAudioNodeOptions const&)
```

Where `DefaultAudioOptions` is just AudioNodeOptions without any defaults. And the setter for channel count as an example could then look like:

```cpp
ExceptionOr<void> AudioNode::intialize_audio_node_options(AudioNodeOptions const& given_options, DefaultAudioNodeOptions const& default_options)
{
    TRY(set_channel_count(given_options.channel_count.value_or(default_options.channel_count));

    // .. etc for the other AudioNode options
}
```

`intialize_audio_node_options` can then be invoked within `construct_impl` with the given AudioOptions, along with the default AudioOptions for that specific audio node.

Reply:
I may not have gotten the implementation of `intialize_audio_node_options` exactly correct to spec, but I think the concept in general should give the required information about user provided vs node default

Reply:
Ah, and one thing I missed in that comment is that the node specific "constraints" on a specific option could be set on the node by virtualizing `set_channel_count` etc